torch/csrc/jit/passes/xnnpack_rewrite.cpp

```diff
@@ -296,13 +296,28 @@ void FoldPrePackingOps(script::Module& m) {
   PrePackingOpsFolder(m, filter_fn, "prepack_folding");
 }
 
-c10::optional<script::Module> optimizeForMobile(const script::Module& m) {
+script::Module optimizeForMobile(
+    const script::Module& m,
+    const std::map<MobileOptimizerType, bool>& whitelist_optimizers) {
   auto cloned_module = m.clone();
   cloned_module.eval();
-  cloned_module = FoldConvBatchNorm2d(cloned_module);
-  insertPrePackedOps(cloned_module);
-  cloned_module = freeze_module(cloned_module);
-  FoldPrePackingOps(cloned_module);
+
+  auto bn_it =
+      whitelist_optimizers.find(MobileOptimizerType::FOLD_CONV_BATCH_NORM);
+  if (bn_it == whitelist_optimizers.end() ||
+      (bn_it != whitelist_optimizers.end() && bn_it->second)) {
+    cloned_module = FoldConvBatchNorm2d(cloned_module);
+  }
+
+  auto prepack_it =
+      whitelist_optimizers.find(MobileOptimizerType::INSERT_FOLD_PREPACK_OPS);
+  if (prepack_it == whitelist_optimizers.end() ||
+      (prepack_it != whitelist_optimizers.end() && prepack_it->second)) {
+    insertPrePackedOps(cloned_module);
+    cloned_module = freeze_module(cloned_module);
+    FoldPrePackingOps(cloned_module);
+  }
+
```

Comment:
This pattern is fine for two or three use cases but when and if the list of options grow we want to switch to something like this:

```
// Psudo code - most probably won't compile without modifications.
struct OpCode2Op {
  MobileOptimizerType opcdoe;
  std::function< script::Module (const script::Module&) > op;
};

const std::vector<OpCode2Op> opcodes2ops = {
  // OP 1
  {
     MobileOptimizerType::FOLD_CONV_BATCH_NORM,
     [](const script::Module& module) {
           return FoldConvBatchNorm2d(cloned_module);
     },
  },
  // OP 2
  {
    ...
  },
  // OP 3 and so on
  ...
};

auto cloned_module = m.clone();
cloned_module.eval();

foreach element in opcodes2ops:
  if element.opcode not in blacklist or is it is but not disabled:
     cloned_module = element->op(cloned_module);
```

Reply:
I like this suggestion. We can have a `const static` instance of OpCode2Op, or call it OptimizationTypes as Ashkan showed?
```
struct OptimizationType {
  MobileOptimizerType type;
  std::function< script::Module (const script::Module&) > apply;
};
```
Also we may need to run optimization in particular order. Meaning if you run batch norm fused after prepacking to insert xnnpack conv/linear then bn folding wont work. So we need a way to capture that as well. std::vector kind of works for it, but it does not capture the intent.

Reply:
Thanks for the suggestion, It looks good. Let me create a struct and attach a function for each pass. @kimishpatel Do we need to cover running optimization in particular order here?

Reply:
Yes the order matters. Therefore I am not sure if doing thing is necessarily good:
```
foreach element in opcodes2ops:
  if element.opcode not in blacklist or is it is but not disabled:
     cloned_module = element->op(cloned_module);
```
@AshkanAliabadi, what do you suggest? I am of the opinion that we shouldn't just loop over the vector. It gets rid of the order information. I suggest we do something like:
```
assuming optimization_blacklist is a set
if (!optimization_blacklist.count(ConvBNFusion)) {
optimization_types.find(ConvBNFusion)->apply();
}
```
Or we can just keep it simple to be honest. 
In fact, while I like the idea, lets do it when we actually need to do it. Right now we dont have that many so we can keep it simple.
```
if (!optimization_blacklist.count(ConvBNFusion)) {
FoldConvBatchNorm2d(module)
}
```

Reply:
I think not looping the list and just checking if it is in blacklist before executing if order matters. @AshkanAliabadi are you good with it?

Reply:
Yes definitely.

Generally speaking anything I say is just a suggestion so please feel free to ignore.  I believe the person writing the code or developing a feature has the best context and understanding of what's going on so please do as what you think makes best sense.  Thanks again.